linode/manager
# feat: [M3-6969] - Add DC-specific pricing to Kubernetes HA

## Description 📝
High Availability (HA) pricing ($60.00/month per cluster for all DCs) is controlled through an environment variable in CI. This PR moves base pricing to a constant in the codebase and allows us to dynamically calculate HA pricing for new DCs.

## Major Changes 🔄
- With the feature flag on:
   - On the Kubernetes Create page, summary/pricing data and the Create Cluster button do not appear in the Checkout Summary bar until a Region selection has been made. This will avoid displaying inaccurate pricing in the checkout bar, since a user could enable HA before choosing a Region. 
   - Prices on enable HA Control Plane radio button on Kubernetes Create page reflect DC-specific HA prices.
      - When no Region is selected, no pricing information is shown.
   - Prices on enable Update to HA modal on Kubernetes Details page reflect DC-specific HA prices.
   - Prices on the Kube Summary section of the Kubernetes Details page reflect DC-specific HA prices.
- HA pricing relies on the pricing multiplier constant (`LKE_HA_PRICE`) rather than the `REACT_APP_LKE_HIGH_AVAILABILITY_PRICE` env var.
- Unit tests are updated and the HA selection that had been previously commented out in e2e tests was restored with the use of a constant for HA. 
- Updates Sao Paulo's price increase percentage. 

## Preview 📷
| Location  | Preview   |
| ------- | ------- |
| Create Cluster | <video src="https://github.com/linode/manager/assets/114685994/ba15ca1b-df84-4d19-baa5-7a38b6aba09b"/> |
| Create Cluster - No Region | ![Screenshot 2023-08-20 at 3 30 43 PM](https://github.com/linode/manager/assets/114685994/cbe6a392-d164-4ba9-83eb-4d2dc11a5156) |
| Kube Cluster Specs | ![Screenshot 2023-08-20 at 3 32 16 PM](https://github.com/linode/manager/assets/114685994/5718c47b-7f76-4c5d-af60-076351f4333b) |
| Kube Upgrade Cluster Dialog | ![Screenshot 2023-08-20 at 3 37 20 PM](https://github.com/linode/manager/assets/114685994/c0bdd032-a38d-441c-9e6a-628e0a2af304) |

## How to test 🧪
1. **How to setup test environment?**
- Check out code locally and enable the DC-Specific Pricing feature flag in the dev settings.

2. **How to verify changes?**
- Verify the e2es and unit tests pass.
- Test the following flows with both the flag on and off. Test Sao Paulo, Jakarta and any region without a price increase.
- When DC-specific pricing flag is *off*, the flows should match production. When *on*, the flows should show the major changes described above and visible in the preview section. 

Flows:
- http://localhost:3000/kubernetes/create
   - Without selecting a Region, toggle between HA selections. Confirm no pricing is displayed in the Checkout Summary.
   - Selecting a Region, confirm the pricing is displayed in the Checkout Summary. Test all DC-specific pricing regions.
   - Select a Plan and confirm the total is correct. (note this total)
   - *Do not* enable HA. (select 'No')
   - Create Cluster.
- http://localhost:3000/kubernetes/clusters/{id}/summary
   - Confirm the total price in the Kube Summary (top of the page) is accurate, matching the total from the Create page.
   - Click the Upgrade to HA button and confirm the copy and the HA price on the dialog is correct. 
   - Submit the form to upgrade to HA and confirm the total price in the Kube Summary has been updated to reflect the addition of HA.

4. **How to run Unit or E2E tests?**
```
yarn test HAControlPlane KubeCheckoutBar kubeUtils
```

```
yarn cy:run -s "cypress/e2e/core/kubernetes/smoke-lke-create.spec.ts cypress/e2e/core/kubernetes/lke-create.spec.ts"
```
